Sources/SwiftSyntaxMacros/SwiftSyntaxMacros.docc/SwiftSyntaxMacros.md

```diff
@@ -1,37 +1,308 @@
+SwiftSyntax contains a `Keyword` enum with all the keywords that can be used in the Swift language. Suppose we need an enum that only contains those keywords that can start type declarations, like `class`, `struct` and `actor` and we need to be able to convert between this type and the `Keyword` type. 
+
+A hand-written implementation could look as follows, which is very repetitive. The goal is to define an `EnumSubset` macro that generates the initializer and the computed property.
+
+```swift
+enum TypeDeclarationKeyword {
+  case `actor`
+  case `class`
+  ...
+
+  init?(_ keyword: Keyword) {
+    switch keyword {
+    case .actor: self = .actor
+    case .class: self = .class
+    ...
+    default: return nil
+    }
+  }
+
+  var keyword: Keyword {
+    switch self {
+    case .actor: return .actor
+    case .class: return .class
+    ...
+    }
+  }
+}
+```
+
+### Declare EnumSubset
+
+Since both the initializer and the computed property are members of `TypeDeclarationKeyword`, the `EnumSubset` macro needs to be an attached member macro. In the following, we will only generate the initializer. Generating the computed property is analogous.
+
+The macro declaration looks as follows.
+
+```swift
+@attached(member, names: named(init))
+public macro EnumSubset<Superset>() = #externalMacro(module: "MyMacroMacros", type: "EnumSubsetMacro")
+```
+
+Compared to the `stringify` macro, `EnumSubset` differs in two ways:
+1. It is declared as an attached member macro and defines the names of the members it introduces. Declaring the introduced names improves the compiler’s performance: When accessing a member that is not declared in the macro’s names, the compiler doesn’t need to try expanding the macro to find it.
+2. `EnumSubset` doesn’t take any argument. Instead, it defines a generic parameter that is used to customize the superset – in the example this is `Keyword`. Using a generic parameter ensures that the superset type exists.
+
+### Implement EnumSubset
+
+A possible implementation of `EnumSubsetMacro` can look as follows. 
+In the following sections, we explore techniques for how to write this code.
+
+```swift
+public enum EnumSubsetMacro: MemberMacro {
+  public static func expansion(
+    of attribute: AttributeSyntax,
+    providingMembersOf declaration: some DeclGroupSyntax,
+    in context: some MacroExpansionContext
+  ) throws -> [DeclSyntax] {
+    guard let enumDecl = declaration.as(EnumDeclSyntax.self) else {
+      // Emit an error here, see section *Emitting Errors*
+      return []
+    }
+
+    // Extract the name of the generic parameter. 
+    // See section *Inspect the SwiftSyntax Tree* how to build this expresion.
+    guard let supersetType = attribute,
+      .attributeName.as(SimpleTypeIdentifierSyntax.self)?
+      .genericArgumentClause?
+      .arguments.first?
+      .argumentType else {
+      // Emit an error here.
```

Comment:
I think it would be good to include emitting the error here to serve as a complete example.

Reply:
Added the error handling to the example.